core/client/client_test.go

```diff
@@ -104,4 +112,135 @@ func TestCompressHistory(t *testing.T) {
+	userID := "fakeuser"
+
+	type request struct {
+		wantStart int64
+		wantSize  int32
+		next      int64
+		items     []*pb.GetEntryResponse
+	}
+
+	for _, tc := range []struct {
+		desc       string
+		start, end int64
+		reqs       []request
+		pageSize   int32
+		wantErr    error
+	}{
+		{
+			desc:    "incomplete",
+			end:     10,
+			reqs:    []request{{}},
+			wantErr: ErrIncomplete,
+		},
+		{
+			desc: "1Item",
+			end:  0,
+			reqs: []request{
+				{items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{0}}},
+				}},
+			},
+		},
+		{
+			desc:  "3Times",
+			start: 0,
+			end:   10,
+			reqs: []request{
+				{wantStart: 0, next: 5, items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{0}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{1}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{2}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{3}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{4}}},
+				}},
+				{wantStart: 5, next: 10, items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{5}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{6}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{7}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{8}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{9}}},
+				}},
+				{wantStart: 10, next: 0, items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{10}}},
+				}},
+			},
+		},
+		{
+			desc:     "pageSize",
+			start:    0,
+			end:      5,
+			pageSize: 3,
+			reqs: []request{
+				// The value of next is opaque to the client.
+				{wantStart: 0, wantSize: 3, next: 1, items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{0}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{1}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{2}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{3}}},
+				}},
+				{wantStart: 1, wantSize: 2, next: 0, items: []*pb.GetEntryResponse{
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{5}}},
+					{Smr: &trillian.SignedMapRoot{MapRoot: []byte{6}}},
+				}},
+			},
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+			s, stop, err := testutil.NewMockKT(ctrl)
+			if err != nil {
+				t.Fatalf("NewMockKT(): %v", err)
+			}
+			defer stop()
+
+			c := Client{
+				Verifier: &fakeVerifier{},
+				cli:      s.Client,
+				pageSize: tc.pageSize,
+			}
+
+			for _, r := range tc.reqs {
+				s.Server.EXPECT().ListEntryHistory(gomock.Any(), matchers.ProtoEqual(&pb.ListEntryHistoryRequest{
+					AppId:    appID,
+					UserId:   userID,
+					Start:    r.wantStart,
+					PageSize: r.wantSize,
+				})).Return(&pb.ListEntryHistoryResponse{
+					NextStart: r.next,
+					Values:    r.items,
+				}, nil)
+			}
+
+			if _, _, err = c.PaginateHistory(ctx, appID, userID, tc.start, tc.end); err != tc.wantErr {
```

Comment:
This method of testing PaginateHistory seems odd, given you could test based on state - i.e. just test PaginateHistory based on its inputs and outputs, rather than checking that it's calling the correct server endpoints with the right arguments....  In fact, by doing the way you currently are explicitly *doesn't* test some of the logic within PaginateHistory outside of what it sends/receives to/from the server.

In addition to this, (and as discussed separately from this PR,) use of gomock (and mocks in general) tends to be discouraged from a readability standpoint, preferring real objects or fakes.

If you were to test PaginateHistory based on inputs and outputs (and compare them to expected outputs) you could just fake the one rpc it calls (ListEntryHistory is it?), which would remove the need for mocking, making the test code much simpler.  (Although incidentally it looks like the PaginateHistory code doesn't call that directly, but instead calls client.VerifiedListHistory, which then calls ListEntryHistory - I couldn't find a test for VerifiedListHistory, so using this method of testing could also be appropriate for that function too.)

This method of testing would be specifically testing the PaginateHistory code, and assuming the underlying functions that it calls work as expected.